### PR TITLE
    orchagent: Adding LagOrch

### PR DESCRIPTION
**Pull request by stcheng**
orchagent: Adding LagOrch

    - Add LagOrch class to listen to APP_LAG_TABLE
    - Four private functions added in lagorch.h:
        addLag, removeLag, addLagMember, removeLagMember
    - Add setPort/removePort functions in portsorch.h

### CODE REVIEW

File: orchagent/lagorch.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        return false;
+    }
+
+    SWSS_LOG_ERROR("Add member %s to LAG %s lid:%llx pid:%llx\n",
+            port.m_alias.c_str(), lag.m_alias.c_str(), lag.m_lag_id, port.m_port_id);
+
+    port.m_type = Port::LAG_MEMBER;
+    port.m_lag_id = lag.m_lag_id;
+    port.m_lag_member_id = lag_member_id;
+    m_portsOrch->setPort(port.m_alias, port);
```
Comment (lguohan) [inline orchagent/lagorch.cpp]: why need to setPort here? looks like you are doing the same thing at line 214?

File: orchagent/lagorch.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        SWSS_LOG_ERROR("Failed to add member %s to LAG %s lid:%llx pid:%llx\n",
+                port.m_alias.c_str(), lag.m_alias.c_str(), lag.m_lag_id, port.m_port_id);
+        return false;
+    }
+
+    SWSS_LOG_ERROR("Add member %s to LAG %s lid:%llx pid:%llx\n",
+            port.m_alias.c_str(), lag.m_alias.c_str(), lag.m_lag_id, port.m_port_id);
+
+    port.m_type = Port::LAG_MEMBER;
```
Comment (lguohan) [inline orchagent/lagorch.cpp]: the Port::LAG_MEMBER and Port::PHY_PORT are different attribute for a physical port. should not be exclusive to each other. A physical port can be both PHY_PORT and LAG_MEMBER.

File: orchagent/lagorch.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    m_portsOrch->setPort(lag, p);
+
+    m_lags[lag] = set<string>();
+
+    return true;
+}
+
+bool LagOrch::removeLag(Port lag)
+{
+    /* Retry when the LAG still has members */
```
Comment (lguohan) [inline orchagent/lagorch.cpp]: need to test in real cases whether these events can be reordered or not.

File: orchagent/lagorch.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                }
+
+                if (addLag(lag_alias))
+                    it = consumer.m_toSync.erase(it);
+                else
+                    it++;
+            }
+            else if (op == DEL_COMMAND)
+            {
+                assert(m_lags.find(lag_alias) != m_lags.end());
```
Comment (lguohan) [inline orchagent/lagorch.cpp]: why do we need to maintain this m_lags as we have lag also stored in m_portList?

File: orchagent/lagorch.cpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                    it = consumer.m_toSync.erase(it);
+                else
+                    it++;
+            }
+            else if (op == DEL_COMMAND)
+            {
+                assert(m_lags[lag_alias].find(port_alias) != m_lags[lag_alias].end());
+                assert(port.m_type == Port::LAG_MEMBER);
+
+                if (removeLagMember(lag, port))
```
Comment (lguohan) [inline orchagent/lagorch.cpp]: check if the port is already a lag member before you do the remove, you need to perform this sanity check in at APP DB level.

File: orchagent/lagorch.cpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                {
+                    SWSS_LOG_ERROR("Duplicate LAG member entry lag:%s port:%s",
+                            lag_alias.c_str(), port_alias.c_str());
+                    it = consumer.m_toSync.erase(it);
+                    continue;
+                }
+
+                assert(port.m_type == Port::PHY_PORT);
+
+                if (addLagMember(lag, port))
```
Comment (lguohan) [inline orchagent/lagorch.cpp]: check if the port is already in a lag or not (cloud be the same lag or a different lag).

**Comment by kcudnik** [comment]
Ack

**Pull request merged**